1121 subsetretrieve observations add gtsmip c3s locations and data

n-aleksandrova opened the pull request:
GTSM reanalysis data from CDS was added to the observations.py to enable retrieving a dataframe of GTSM output locations and downloading GTSM-ERA5 data from CDS.

The GTSM obs points are read from the [.csv file in the GTSM-ERA5 repository](https://github.com/VU-IVM/gtsm3-era5-nrt/blob/main/04_supplementary_data/GTSM_output_locations_list.csv). Please note that the station names are not always identical to the station names handled by the model, because some names included special characters which I have removed in postprocessing (this was causing issues when using the station name in filenames when saving individual observations) (example: 'currents_l�staviken' vs. 'currents_lstaviken').

The GTSM data cannot be subset by station when downloading from CDS, therefore the data is downloaded once to a cache folder, and then accessed to retrieve data for individual stations.

Possible further improvements:
- Currently this code retrieves hourly timeseries from CDS. We can optionally add 10-min timeseries as well, this is not done yet.
- For any area, there are many more GTSM stations than stations from other sources (x10 more). To mitigate this a bit in the [subset/retrieve notebook](https://github.com/Deltares/dfm_tools/blob/main/docs/notebooks/subset_retrieve_sealevel_observations.ipynb), I am subsetting the GTSM dataset to exclude gridded points in open sea (keeping only points located close to the coast).
- The large number of GTSM stations results in a very long legend for the figure with the timeseries. For now I have commented the legend out, but we can consider to not plot GTSM data in this overview. 
- it seems unnecessary to show GTSM data in the overview created by dfmt.ssh_netcdf_overview() because there will always be data available for any period in 1950-2024. But there is no simple way to exclude this data, so I left it as it is.

veenstrajelmer commented on dfm_tools/observations.py:
```diff
@@ -849,6 +871,72 @@ def rwsddl_ssh_retrieve_data(row, dir_output, time_min, time_max):
+    dir_cache = get_dir_testdata()
+
+    # Define timeframe of the request as lists of years and months
+    time_min_dt = datetime.strptime(time_min, '%Y-%m-%d')
+    time_max_dt = datetime.strptime(time_max, '%Y-%m-%d')
```
Neater is `pd.Timestamp(time_min)`, but you don't need it when using `pd.data_range` or `pd.period_range`, since pandas just takes strings. The user should also be able to supply a time min/max of type datetime/pandas/str. Your code will fail if the argument is not a string.

n-aleksandrova replied:
Changed to pd.Timestamp.

veenstrajelmer commented on dfm_tools/observations.py:
```diff
@@ -849,6 +871,72 @@ def rwsddl_ssh_retrieve_data(row, dir_output, time_min, time_max):
+                os.remove(tmp_zipfile)
+
+    # open dataset
+    ds = xr.open_mfdataset(os.path.join(dir_cache, f'reanalysis_waterlevel_hourly_*_v2.nc'))
```
- include some arguments in open_mfdataset() to make it faster and more robust (see our teams chat of last week)
- sonarcloud complains about you using an f-string without arguments, so you can remove the f here

veenstrajelmer commented on tests/test_observations.py:
```diff
@@ -38,12 +38,12 @@ def test_ssh_catalog_subset(source):
         assert field in ssc_catalog_gpd.columns
-    if source not in ["ssc", "psmsl-gnssir", "rwsddl"]:
+    if source not in ["ssc", "psmsl-gnssir", "rwsddl","gtsm-era5-cds"]:
         assert "time_ndays" in ssc_catalog_gpd.columns
     assert ssc_catalog_gpd.crs.to_string()=='EPSG:4326'
     
     # do actual subset
-    source_list_notime = ["ssc","rwsddl"]
+    source_list_notime = ["ssc","rwsddl","gtsm-era5-cds"]
```
Should be removed from these lists, since the time extent is known

veenstrajelmer commented on dfm_tools/observations.py:
```diff
@@ -849,6 +871,72 @@ def rwsddl_ssh_retrieve_data(row, dir_output, time_min, time_max):
+    # Retrieve data via an API request and extract archive - for files not found in the cache
+    for yy,year in enumerate(years):
+        for mm,month in enumerate(months[yy]):
+            filename = os.path.join(dir_cache, f'reanalysis_waterlevel_hourly_{year}_{month:02}_v2.nc')
```
is "v2" also present in the original files?

n-aleksandrova replied:
Yes, this is the filename that is downloaded from CDS for all files (1950-2024).

veenstrajelmer commented on dfm_tools/observations.py:
```diff
@@ -849,6 +871,72 @@ def rwsddl_ssh_retrieve_data(row, dir_output, time_min, time_max):
+    if time_min is None or time_max is None:
+        raise ValueError("cannot supply None for 'time_min' or 'time_max' to 'gtsm_era5_cds_retrieve_data()'")
+    
+    dir_cache = get_dir_testdata()
```
Please add a subdirectory so the cachedir does not get flooded with >100 files.

n-aleksandrova replied:
Good point, I added a subfolder.

veenstrajelmer reviewed (changes requested):
Thanks for this valuable contribution, including the very clear description with discussion points. Very neat! It looks like it will work well (have not tested it yet, I will do that after your changes), but I do have added some suggestions to specific parts of the code. Please let me know of any of these review comments are annoying to you. I realize I am micro-managing (for a good cause, but still), so you can drop any of the tasks back at my desk.

For the notebook (too large diff to comment inline):
- maybe skip downloading (and plotting subset) just like gesla, this avoids all issues you encountered. But of course do keep them in the global overview. I am also planning to update the modelbuilder notebook to include GTSM as model obspoints, so there will be exposure nevertheless.
- remove the new gtsm source from the TODO comment again (see other review points)

n-aleksandrova commented on dfm_tools/observations.py:
```diff
@@ -549,6 +552,25 @@ def rwsddl_ssh_read_catalog(meta_dict=None):
+    
+    stat_names = "gtsm-era5-" + station_list_gpd['station_id'].astype(str) + "-" + station_list_gpd['station_name']
+    station_list_gpd["station_name_unique"] = stat_names
+    station_list_gpd["country"] = 'global'
```
Changed country to ""
Added start time, end time and number of days to the dataframe.

n-aleksandrova commented on dfm_tools/observations.py:
```diff
@@ -22,6 +22,9 @@
 import fiona
 import copernicusmarine
+import cdsapi
+from datetime import datetime
```
Removed datetime, using pd.Timestamp instead

n-aleksandrova commented on dfm_tools/observations.py:
```diff
@@ -849,6 +871,72 @@ def rwsddl_ssh_retrieve_data(row, dir_output, time_min, time_max):
+    The function checks if the files have already been downloaded to cache
+    """
+    
+    if time_min is None or time_max is None:
+        raise ValueError("cannot supply None for 'time_min' or 'time_max' to 'gtsm_era5_cds_retrieve_data()'")
```
Removing this from the code. However to account for the situation where time_min or time_max is None, we probably need an exception to get time_min or time_max from the dataframe?
The full dataset will be very large, I don't think that should be the default option to download the full 75-year dataset.

n-aleksandrova commented on dfm_tools/observations.py:
```diff
@@ -849,6 +871,72 @@ def rwsddl_ssh_retrieve_data(row, dir_output, time_min, time_max):
+    time_min_dt = datetime.strptime(time_min, '%Y-%m-%d')
+    time_max_dt = datetime.strptime(time_max, '%Y-%m-%d')
+    if time_max_dt.year == time_min_dt.year: # dates within one year
+        years = [time_max_dt.year]
+        months = [list(range(time_min_dt.month, time_max_dt.month+1))]
+    elif (time_max_dt.year - time_min_dt.year) == 1: # dates across the year boundary
+        years = [time_min_dt.year, time_max_dt.year]
+        months = [list(range(time_min_dt.month, 13)), list(range(1, time_max_dt.month))]
+    elif (time_max_dt.year - time_min_dt.year) > 1: # dates span multiple years
+        years = list(range(time_min_dt.year, time_max_dt.year+1))
+        months = [list(range(time_min_dt.month, 13))]
+        for yy in years[1:-1]:
+            months.append(list(range(1, 13)))
+        months.append(list(range(1, time_max_dt.month)))
+
+    # Retrieve data via an API request and extract archive - for files not found in the cache
+    for yy,year in enumerate(years):
+        for mm,month in enumerate(months[yy]):
+            filename = os.path.join(dir_cache, f'reanalysis_waterlevel_hourly_{year}_{month:02}_v2.nc')
+            if not os.path.isfile(filename):
+                print('... retrieving GTSM-ERA5 data for %i, month %i' % (year,month))    
+                tmp_zipfile = os.path.join(dir_cache, f'gtsm_era5_{year}_{month:02}.zip')
+
+                # Make connection with CDS via API (assumes URL and KEY already stored in .cdsapirc file)
+                c = cdsapi.Client() 
+                c.retrieve(
+                    'sis-water-level-change-timeseries-cmip6',
+                    {
+                        'variable': 'total_water_level',
+                        'experiment': 'reanalysis',
+                        'temporal_aggregation': 'hourly',
+                        'year': str(year),
+                        'month': [str(x).zfill(2) for x in months[yy]],
+                        'format': 'zip',
+                    }, 
+                    tmp_zipfile)
+            
+                with ZipFile(tmp_zipfile, 'r') as zip_ref:
+                    zip_ref.extractall(dir_cache)
+                os.remove(tmp_zipfile)
```
Thanks for these suggestions! I have rewritten the code using pd.time_range and simplified the for-loop. I think it looks much simpler now.

n-aleksandrova commented:
I have submitted a new version of the scripts, with adjustments based on the discussion we had.

In summary:
- Source for GTSM was renamed to `gtsm3-era5-cds`.
- Added option to retrieve surge timeseries, and choose between hourly and 10-min time resolutions for GTSM. **To do:** currently all retrieved TS are saved to files with station name in its name and 'waterlevel' variable, we need to enable distinguishing between the waterlevel and surge timeseries. This is however a 'special case' compared to other observation sources. To be discussed how to best do that without adding too many exceptions to the code. 
- Fixed other minor issues related to formatting and visualizations.
- ModelBuilder notebook example now includes the retrieval and visualization of the obs points based on GTSM. **Note:** I did not yet test if the model runs correctly with these new obs points.

veenstrajelmer commented:
Some minor things added:
- removed support for surge again. It made the code complex and impossible to distuinguish in the netcdf files as you also noted. Furthermore, there is no request for this yet.
- whatsnew: linked to PR instead of issue
- added testcase to cover time_min/time_max =None and invalid time_freq
- suppressed the CMEMS logging in the notebook
